File: src/hexa_display/test/test_expression_policy.py
import pytest

from hexa_display.expression_policy import (
    DEFAULT_EXPRESSION_MAP,
    IDLE_TARGET,
    BatteryMonitor,
    DisplayTarget,
    PolicyConfig,
    PolicyInputs,
    decide,
    quantize_axis,
    select_face_animation,
)
from hexa_display.face_animation import FACE_ANIMATIONS
from hexa_display.protocol import Expression, Gaze

CONFIG = PolicyConfig(expression_map=dict(DEFAULT_EXPRESSION_MAP))


def make_inputs(**kwargs) -> PolicyInputs:
    defaults = dict(
        gait_state="stand",
        vx=0.0,
        vy=0.0,
        wz=0.0,
        animation_mode="",
        roll=0.0,
        pitch=0.0,
        yaw=0.0,
        battery_low=False,
        battery_critical=False,
    )
    defaults.update(kwargs)
    return PolicyInputs(**defaults)


@pytest.mark.parametrize("state,expression", list(DEFAULT_EXPRESSION_MAP.items()))
def test_gait_state_map_defaults(state, expression):
    target = decide(make_inputs(gait_state=state), CONFIG, IDLE_TARGET)
    assert target.expression == expression


def test_unknown_and_unseen_state_are_neutral():
    assert (
        decide(make_inputs(gait_state="warp"), CONFIG, IDLE_TARGET).expression
        == Expression.NEUTRAL
    )
    assert (
        decide(make_inputs(gait_state=None), CONFIG, IDLE_TARGET).expression
        == Expression.NEUTRAL
    )


def test_animation_mode_wins_over_gait_state():
    target = decide(
        make_inputs(gait_state="gait", animation_mode="body_roll_3d"),
        CONFIG,
        IDLE_TARGET,
    )
    assert target.expression == Expression.WOOZY


def test_battery_warning_only_when_idle():
    idle = decide(
        make_inputs(gait_state="stand", battery_low=True), CONFIG, IDLE_TARGET
    )
    assert idle.expression == Expression.SLEEPY
    walking = decide(
        make_inputs(gait_state="gait", vx=0.05, battery_low=True),
        CONFIG,
        IDLE_TARGET,
    )
    assert walking.expression == Expression.HAPPY
    animating = decide(
        make_inputs(battery_low=True, animation_mode="vertical_body_roll"),
        CONFIG,
        IDLE_TARGET,
    )
    assert animating.expression == Expression.WOOZY


def test_battery_critical_overrides_everything_and_centers_gaze():
    target = decide(
        make_inputs(
            gait_state="gait",
            vx=0.1,
            animation_mode="body_roll_3d",
            battery_low=True,
            battery_critical=True,
        ),
        CONFIG,
        IDLE_TARGET,
    )
    assert target == DisplayTarget(expression=Expression.DEAD, gaze=Gaze.CENTER)


@pytest.mark.parametrize(
    "vy,wz,gaze",
    [
        (0.1, 0.0, Gaze.LEFT),
        (-0.1, 0.0, Gaze.RIGHT),
        (0.0, 0.5, Gaze.LEFT),
        (0.0, -0.5, Gaze.RIGHT),
    ],
)
def test_gaze_follows_cmd_vel_horizontally(vy, wz, gaze):
    target = decide(
        make_inputs(gait_state="gait", vx=0.05, vy=vy, wz=wz),
        CONFIG,
        IDLE_TARGET,
    )
    assert target.gaze == gaze


def test_forward_backward_motion_keeps_gaze_level():
    forward = decide(make_inputs(gait_state="gait", vx=0.1), CONFIG, IDLE_TARGET)
    assert forward.gaze == Gaze.CENTER
    backward = decide(
        make_inputs(gait_state="gait", vx=-0.1), CONFIG, IDLE_TARGET
    )
    assert backward.gaze == Gaze.CENTER


def test_pitch_drives_vertical_gaze_while_gait_active():
    up = decide(
        make_inputs(gait_state="gait", vx=0.1, pitch=-0.2), CONFIG, IDLE_TARGET
    )
    assert up.gaze == Gaze.UP
    down_left = decide(
        make_inputs(gait_state="gait", vx=0.1, vy=0.1, pitch=0.2),
        CONFIG,
        IDLE_TARGET,
    )
    assert down_left.gaze == Gaze.DOWN_LEFT


def test_gaze_below_deadband_is_center():
    # 0.001 / vy_max 0.1 = 0.01 normalized, far under the 0.15 deadband.
    target = decide(
        make_inputs(gait_state="gait", vy=0.001), CONFIG, IDLE_TARGET
    )
    assert target.gaze == Gaze.CENTER


def test_gaze_hysteresis_holds_direction_in_exit_band():
    prev = DisplayTarget(expression=Expression.HAPPY, gaze=Gaze.LEFT)
    # 0.012 / 0.1 = 0.12 normalized: under deadband 0.15 but above the
    # exit level 0.15 * 0.6 = 0.09, so LEFT is held.
    held = decide(make_inputs(gait_state="gait", vy=0.012), CONFIG, prev)
    assert held.gaze == Gaze.LEFT
    # Fresh entry from CENTER at the same value stays CENTER.
    fresh = decide(make_inputs(gait_state="gait", vy=0.012), CONFIG, IDLE_TARGET)
    assert fresh.gaze == Gaze.CENTER
    # Below the exit level the held direction releases.
    released = decide(make_inputs(gait_state="gait", vy=0.005), CONFIG, prev)
    assert released.gaze == Gaze.CENTER


def test_pose_mode_gaze_follows_tilt():
    # Negative pitch = nose up → gaze UP; positive yaw = left → LEFT.
    up = decide(make_inputs(pitch=-0.2), CONFIG, IDLE_TARGET)
    assert up.gaze == Gaze.UP
    down = decide(make_inputs(pitch=0.2), CONFIG, IDLE_TARGET)
    assert down.gaze == Gaze.DOWN
    left = decide(make_inputs(yaw=0.2), CONFIG, IDLE_TARGET)
    assert left.gaze == Gaze.LEFT
    right = decide(make_inputs(roll=0.2), CONFIG, IDLE_TARGET)
    assert right.gaze == Gaze.RIGHT
    level = decide(make_inputs(), CONFIG, IDLE_TARGET)
    assert level.gaze == Gaze.CENTER


def test_breathing_selected_while_waiting_for_the_stack():
    assert (
        select_face_animation(make_inputs(gait_state=None), CONFIG)
        == "breathing"
    )


def test_idling_selected_when_standing_idle_and_level():
    assert (
        select_face_animation(make_inputs(gait_state="stand"), CONFIG)
        == "idling"
    )


@pytest.mark.parametrize(
    "inputs",
    [
        make_inputs(gait_state="gait", vx=0.1),  # walking
        make_inputs(gait_state="stand", wz=0.3),  # turning in place
        make_inputs(gait_state="stand", pitch=0.2),  # pose mode tilt
        make_inputs(gait_state="stand", yaw=0.2),
        make_inputs(gait_state="folded"),  # sleepy states stay still
        make_inputs(gait_state="paused"),
        make_inputs(gait_state="stand", animation_mode="body_roll_3d"),
        make_inputs(gait_state="stand", battery_low=True),
        make_inputs(gait_state=None, battery_critical=True),
    ],
)
def test_no_face_animation_when_busy_or_warning(inputs):
    assert select_face_animation(inputs, CONFIG) is None


def test_selected_face_animations_exist_in_registry():
    assert select_face_animation(make_inputs(gait_state=None), CONFIG) in (
        FACE_ANIMATIONS
    )
    assert select_face_animation(make_inputs(gait_state="stand"), CONFIG) in (
        FACE_ANIMATIONS
    )


def test_quantize_axis_basic_and_hysteresis():
    assert quantize_axis(0.2, 0, 0.15, 0.6) == 1
    assert quantize_axis(-0.2, 0, 0.15, 0.6) == -1
    assert quantize_axis(0.1, 0, 0.15, 0.6) == 0
    # In the hold band only the matching previous sign is held.
    assert quantize_axis(0.1, 1, 0.15, 0.6) == 1
    assert quantize_axis(0.1, -1, 0.15, 0.6) == 0
    assert quantize_axis(0.05, 1, 0.15, 0.6) == 0
    # Hard sign flip switches without passing through 0.
    assert quantize_axis(-0.2, 1, 0.15, 0.6) == -1


def test_battery_monitor_disabled_by_zero_thresholds():
    monitor = BatteryMonitor(warning_v=0.0, critical_v=0.0)
    assert monitor.update(0.1, 0.0) == (False, False)
    assert monitor.update(0.1, 100.0) == (False, False)


def test_battery_monitor_hold_time():
    monitor = BatteryMonitor(warning_v=7.0, critical_v=6.4, hold_s=3.0)
    assert monitor.update(6.8, 0.0) == (False, False)  # below, hold running
    assert monitor.update(6.8, 2.9) == (False, False)
    assert monitor.update(6.8, 3.0) == (True, False)
    # A dip that recovers before the hold expires never trips.
    monitor2 = BatteryMonitor(warning_v=7.0, critical_v=6.4, hold_s=3.0)
    assert monitor2.update(6.8, 0.0) == (False, False)
    assert monitor2.update(7.5, 1.0) == (False, False)
    assert monitor2.update(6.8, 2.0) == (False, False)  # hold restarted
    assert monitor2.update(6.8, 4.9) == (False, False)
    assert monitor2.update(6.8, 5.0) == (True, False)


def test_battery_monitor_hysteresis_on_recovery():
    monitor = BatteryMonitor(
        warning_v=7.0, critical_v=0.0, hysteresis_v=0.3, hold_s=0.0
    )
    assert monitor.update(6.9, 0.0) == (True, False)
    # Inside the hysteresis band the flag stays raised.
    assert monitor.update(7.2, 1.0) == (True, False)
    # Above threshold + hysteresis it clears immediately.
    assert monitor.update(7.4, 2.0) == (False, False)


def test_battery_monitor_critical_independent_of_warning():
    monitor = BatteryMonitor(warning_v=7.0, critical_v=6.4, hold_s=0.0)
    assert monitor.update(6.9, 0.0) == (True, False)
    assert monitor.update(6.3, 1.0) == (True, True)

File: src/hexa_display/test/test_expression_policy.cpp
// gtest port of the hexa_display expression_policy pytest suite. Exercises the
// pure decision policy directly — no rclcpp, no device.

#include <optional>
#include <string>

#include <gtest/gtest.h>

#include "Expression.h"
#include "expression_policy.hpp"
#include "face_animation.hpp"

using namespace hexa::display;

namespace {

PolicyConfig makeConfig() {
    PolicyConfig config;
    config.expression_map = defaultExpressionMap();
    return config;
}

// Defaults mirror the pytest make_inputs(): standing, still, level, healthy.
PolicyInputs makeInputs() {
    PolicyInputs in;
    in.gait_state = "stand";
    return in;
}

const PolicyConfig CONFIG = makeConfig();

}  // namespace

TEST(ExpressionPolicy, GaitStateMapDefaults) {
    for (const auto& [state, expression] : defaultExpressionMap()) {
        auto in = makeInputs();
        in.gait_state = state;
        EXPECT_EQ(decide(in, CONFIG, kIdleTarget).expression, expression) << state;
    }
}

TEST(ExpressionPolicy, UnknownStateIsNeutralUnseenStateIsSleepy) {
    auto warp = makeInputs();
    warp.gait_state = "warp";
    EXPECT_EQ(decide(warp, CONFIG, kIdleTarget).expression, Expression::NEUTRAL);

    // The robot boots folded: until /gait/state is heard the eyes stay asleep.
    auto none = makeInputs();
    none.gait_state = std::nullopt;
    EXPECT_EQ(decide(none, CONFIG, kIdleTarget).expression, Expression::SLEEPY);
}

TEST(ExpressionPolicy, AnimationModeWinsOverGaitState) {
    auto in = makeInputs();
    in.gait_state = "gait";
    in.animation_mode = "body_roll_3d";
    EXPECT_EQ(decide(in, CONFIG, kIdleTarget).expression, Expression::WOOZY);
}

TEST(ExpressionPolicy, BatteryWarningOnlyWhenIdle) {
    auto idle = makeInputs();
    idle.gait_state = "stand";
    idle.battery_low = true;
    EXPECT_EQ(decide(idle, CONFIG, kIdleTarget).expression, Expression::SLEEPY);

    auto walking = makeInputs();
    walking.gait_state = "gait";
    walking.vx = 0.05;
    walking.battery_low = true;
    EXPECT_EQ(decide(walking, CONFIG, kIdleTarget).expression, Expression::NEUTRAL);

    auto animating = makeInputs();
    animating.battery_low = true;
    animating.animation_mode = "vertical_body_roll";
    EXPECT_EQ(decide(animating, CONFIG, kIdleTarget).expression, Expression::WOOZY);
}

TEST(ExpressionPolicy, BatteryCriticalOverridesEverythingAndCentersGaze) {
    auto in = makeInputs();
    in.gait_state = "gait";
    in.vx = 0.1;
    in.animation_mode = "body_roll_3d";
    in.battery_low = true;
    in.battery_critical = true;
    EXPECT_EQ(decide(in, CONFIG, kIdleTarget),
              (DisplayTarget{Expression::DEAD, GazeDirection::CENTER}));
}

TEST(ExpressionPolicy, GazeFollowsCmdVelHorizontally) {
    struct Case {
        double vy, wz;
        GazeDirection gaze;
    };
    const Case cases[] = {
        {0.1, 0.0, GazeDirection::LEFT},
        {-0.1, 0.0, GazeDirection::RIGHT},
        {0.0, 0.5, GazeDirection::LEFT},
        {0.0, -0.5, GazeDirection::RIGHT},
    };
    for (const auto& c : cases) {
        auto in = makeInputs();
        in.gait_state = "gait";
        in.vx = 0.05;
        in.vy = c.vy;
        in.wz = c.wz;
        EXPECT_EQ(decide(in, CONFIG, kIdleTarget).gaze, c.gaze)
            << "vy=" << c.vy << " wz=" << c.wz;
    }
}

TEST(ExpressionPolicy, ForwardBackwardMotionKeepsGazeLevel) {
    auto fwd = makeInputs();
    fwd.gait_state = "gait";
    fwd.vx = 0.1;
    EXPECT_EQ(decide(fwd, CONFIG, kIdleTarget).gaze, GazeDirection::CENTER);

    auto bwd = makeInputs();
    bwd.gait_state = "gait";
    bwd.vx = -0.1;
    EXPECT_EQ(decide(bwd, CONFIG, kIdleTarget).gaze, GazeDirection::CENTER);
}

TEST(ExpressionPolicy, PitchDrivesVerticalGazeWhileGaitActive) {
    auto up = makeInputs();
    up.gait_state = "gait";
    up.vx = 0.1;
    up.pitch = -0.2;
    EXPECT_EQ(decide(up, CONFIG, kIdleTarget).gaze, GazeDirection::UP);

    auto down_left = makeInputs();
    down_left.gait_state = "gait";
    down_left.vx = 0.1;
    down_left.vy = 0.1;
    down_left.pitch = 0.2;
    EXPECT_EQ(decide(down_left, CONFIG, kIdleTarget).gaze, GazeDirection::DOWN_LEFT);
}

TEST(ExpressionPolicy, GazeBelowDeadbandIsCenter) {
    auto in = makeInputs();
    in.gait_state = "gait";
    in.vy = 0.001;
    EXPECT_EQ(decide(in, CONFIG, kIdleTarget).gaze, GazeDirection::CENTER);
}

TEST(ExpressionPolicy, GazeHysteresisHoldsDirectionInExitBand) {
    const DisplayTarget prev{Expression::HAPPY, GazeDirection::LEFT};

    auto held = makeInputs();
    held.gait_state = "gait";
    held.vy = 0.012;  // under deadband 0.15 but above exit level 0.09 -> held
    EXPECT_EQ(decide(held, CONFIG, prev).gaze, GazeDirection::LEFT);

    auto fresh = makeInputs();
    fresh.gait_state = "gait";
    fresh.vy = 0.012;
    EXPECT_EQ(decide(fresh, CONFIG, kIdleTarget).gaze, GazeDirection::CENTER);

    auto released = makeInputs();
    released.gait_state = "gait";
    released.vy = 0.005;  // below exit level -> released
    EXPECT_EQ(decide(released, CONFIG, prev).gaze, GazeDirection::CENTER);
}

TEST(ExpressionPolicy, PoseModeGazeFollowsTilt) {
    auto up = makeInputs();
    up.pitch = -0.2;
    EXPECT_EQ(decide(up, CONFIG, kIdleTarget).gaze, GazeDirection::UP);

    auto down = makeInputs();
    down.pitch = 0.2;
    EXPECT_EQ(decide(down, CONFIG, kIdleTarget).gaze, GazeDirection::DOWN);

    auto left = makeInputs();
    left.yaw = 0.2;
    EXPECT_EQ(decide(left, CONFIG, kIdleTarget).gaze, GazeDirection::LEFT);

    auto right = makeInputs();
    right.roll = 0.2;
    EXPECT_EQ(decide(right, CONFIG, kIdleTarget).gaze, GazeDirection::RIGHT);

    EXPECT_EQ(decide(makeInputs(), CONFIG, kIdleTarget).gaze, GazeDirection::CENTER);
}

// Pose mode is where the face answers the sticks: the gait state sits at
// "stand" the whole time the operator poses the body, so the map alone would
// leave the eyes NEUTRAL throughout.
TEST(ExpressionPolicy, PostureSticksDriveExpressionInPoseMode) {
    auto still = makeInputs();
    EXPECT_EQ(decide(still, CONFIG, kIdleTarget).expression, Expression::NEUTRAL);

    auto tilt = makeInputs();  // left stick: body tilt
    tilt.pitch = 0.05;
    EXPECT_EQ(decide(tilt, CONFIG, kIdleTarget).expression, Expression::HAPPY);

    auto roll = makeInputs();
    roll.roll = -0.05;
    EXPECT_EQ(decide(roll, CONFIG, kIdleTarget).expression, Expression::HAPPY);

    auto shift = makeInputs();  // right stick: body translation
    shift.y = 0.02;
    EXPECT_EQ(decide(shift, CONFIG, kIdleTarget).expression, Expression::LOVE);

    auto both = makeInputs();
    both.pitch = 0.05;
    both.x = -0.02;
    EXPECT_EQ(decide(both, CONFIG, kIdleTarget).expression, Expression::ANGRY);
}

// Yaw is L1/R1 and height is the D-pad — neither is a stick, so neither moves
// the expression off the gait state's.
TEST(ExpressionPolicy, YawAloneIsNotAPostureStick) {
    auto in = makeInputs();
    in.yaw = 0.2;
    EXPECT_EQ(decide(in, CONFIG, kIdleTarget).expression, Expression::NEUTRAL);
}

// Walking with a posed body (a recorded posture baseline bleeds through into
// gait mode) stays on the gait state's expression: this only applies in pose
// mode, where the sticks are what moved the body in the first place.
TEST(ExpressionPolicy, PostureExpressionOnlyInPoseMode) {
    auto walking = makeInputs();
    walking.gait_state = "gait";
    walking.vx = 0.05;
    walking.pitch = 0.05;
    walking.y = 0.02;
    EXPECT_EQ(decide(walking, CONFIG, kIdleTarget).expression, Expression::NEUTRAL);

    auto folded = makeInputs();
    folded.gait_state = "folded";
    folded.pitch = 0.05;
    folded.y = 0.02;
    EXPECT_EQ(decide(folded, CONFIG, kIdleTarget).expression, Expression::SLEEPY);
}

TEST(ExpressionPolicy, PostureStickHysteresisHoldsExpressionInExitBand) {
    auto held = makeInputs();
    held.pitch = 0.02;  // under the 0.03 threshold, above the 0.018 exit level
    const DisplayTarget prev{Expression::HAPPY, GazeDirection::CENTER};
    EXPECT_EQ(decide(held, CONFIG, prev).expression, Expression::HAPPY);
    EXPECT_EQ(decide(held, CONFIG, kIdleTarget).expression, Expression::NEUTRAL);

    auto released = makeInputs();
    released.pitch = 0.01;
    EXPECT_EQ(decide(released, CONFIG, prev).expression, Expression::NEUTRAL);

    // Each stick keeps its own band: the shift stick, still held, survives the
    // tilt stick's release.
    auto one_of_two = makeInputs();
    one_of_two.pitch = 0.01;
    one_of_two.x = 0.02;
    const DisplayTarget both{Expression::ANGRY, GazeDirection::CENTER};
    EXPECT_EQ(decide(one_of_two, CONFIG, both).expression, Expression::LOVE);
}

TEST(ExpressionPolicy, PostureStickThresholdOfZeroDisablesTheStick) {
    PolicyConfig config = CONFIG;
    config.posture_shift_threshold_m = 0.0;
    auto shift = makeInputs();
    shift.y = 0.035;  // full deflection
    EXPECT_EQ(decide(shift, config, kIdleTarget).expression, Expression::NEUTRAL);
}

TEST(ExpressionPolicy, BatteryAndBusyOutrankPostureSticks) {
    auto in = makeInputs();
    in.pitch = 0.05;
    in.x = -0.02;

    auto low = in;
    low.battery_low = true;
    EXPECT_EQ(decide(low, CONFIG, kIdleTarget).expression, Expression::SLEEPY);

    auto animating = in;
    animating.animation_mode = "body_roll_3d";
    EXPECT_EQ(decide(animating, CONFIG, kIdleTarget).expression, Expression::WOOZY);

    auto busy = in;
    busy.busy = true;
    EXPECT_EQ(decide(busy, CONFIG, kIdleTarget).expression, Expression::SCANNING);
}

TEST(ExpressionPolicy, PostureSticksSuppressIdling) {
    auto shift = makeInputs();
    shift.y = 0.02;  // level, so only the shift stick keeps idling away
    EXPECT_FALSE(selectFaceAnimation(shift, CONFIG).has_value());

    auto tilt = makeInputs();
    tilt.pitch = 0.05;  // under the gaze's 0.08 level threshold, still a stick
    EXPECT_FALSE(selectFaceAnimation(tilt, CONFIG).has_value());
}

TEST(ExpressionPolicy, ToScreenGazeFlipsHorizontalOnly) {
    EXPECT_EQ(toScreenGaze(GazeDirection::LEFT), GazeDirection::RIGHT);
    EXPECT_EQ(toScreenGaze(GazeDirection::RIGHT), GazeDirection::LEFT);
    EXPECT_EQ(toScreenGaze(GazeDirection::UP_LEFT), GazeDirection::UP_RIGHT);
    EXPECT_EQ(toScreenGaze(GazeDirection::DOWN_RIGHT), GazeDirection::DOWN_LEFT);
    EXPECT_EQ(toScreenGaze(GazeDirection::UP), GazeDirection::UP);
    EXPECT_EQ(toScreenGaze(GazeDirection::DOWN), GazeDirection::DOWN);
    EXPECT_EQ(toScreenGaze(GazeDirection::CENTER), GazeDirection::CENTER);

    for (int i = 0; i < static_cast<int>(GazeDirection::_COUNT); ++i) {
        const auto g = static_cast<GazeDirection>(i);
        EXPECT_EQ(toScreenGaze(toScreenGaze(g)), g) << "gaze " << gazeName(g);
    }
}

TEST(ExpressionPolicy, BreathingSelectedWhileSleeping) {
    // The sleeping face breathes: both before the stack is up and once folded.
    auto waiting = makeInputs();
    waiting.gait_state = std::nullopt;

    auto folded = makeInputs();
    folded.gait_state = "folded";

    for (const auto& in : {waiting, folded}) {
        auto sel = selectFaceAnimation(in, CONFIG);
        ASSERT_TRUE(sel.has_value());
        EXPECT_EQ(*sel, "breathing");
    }
}

TEST(ExpressionPolicy, IdlingSelectedWhenStandingIdleAndLevel) {
    auto sel = selectFaceAnimation(makeInputs(), CONFIG);
    ASSERT_TRUE(sel.has_value());
    EXPECT_EQ(*sel, "idling");
}

TEST(ExpressionPolicy, NoFaceAnimationWhenBusyOrWarning) {
    auto walking = makeInputs();
    walking.gait_state = "gait";
    walking.vx = 0.1;

    auto turning = makeInputs();
    turning.wz = 0.3;

    auto pose_tilt = makeInputs();
    pose_tilt.pitch = 0.2;

    auto pose_yaw = makeInputs();
    pose_yaw.yaw = 0.2;

    auto folding = makeInputs();
    folding.gait_state = "folding";

    auto settling = makeInputs();
    settling.gait_state = "settling";

    auto animating = makeInputs();
    animating.animation_mode = "body_roll_3d";

    auto low = makeInputs();
    low.battery_low = true;

    auto critical = makeInputs();
    critical.gait_state = std::nullopt;
    critical.battery_critical = true;

    for (const auto& in : {walking, turning, pose_tilt, pose_yaw, folding,
                           settling, animating, low, critical}) {
        EXPECT_FALSE(selectFaceAnimation(in, CONFIG).has_value());
    }
}

TEST(ExpressionPolicy, SelectedFaceAnimationsExistInRegistry) {
    auto waiting = makeInputs();
    waiting.gait_state = std::nullopt;
    auto a = selectFaceAnimation(waiting, CONFIG);
    ASSERT_TRUE(a.has_value());
    EXPECT_NE(findFaceAnimation(*a), nullptr);

    auto b = selectFaceAnimation(makeInputs(), CONFIG);
    ASSERT_TRUE(b.has_value());
    EXPECT_NE(findFaceAnimation(*b), nullptr);
}

TEST(ExpressionPolicy, QuantizeAxisBasicAndHysteresis) {
    EXPECT_EQ(quantizeAxis(0.2, 0, 0.15, 0.6), 1);
    EXPECT_EQ(quantizeAxis(-0.2, 0, 0.15, 0.6), -1);
    EXPECT_EQ(quantizeAxis(0.1, 0, 0.15, 0.6), 0);
    // In the hold band only the matching previous sign is held.
    EXPECT_EQ(quantizeAxis(0.1, 1, 0.15, 0.6), 1);
    EXPECT_EQ(quantizeAxis(0.1, -1, 0.15, 0.6), 0);
    EXPECT_EQ(quantizeAxis(0.05, 1, 0.15, 0.6), 0);
    // Hard sign flip switches without passing through 0.
    EXPECT_EQ(quantizeAxis(-0.2, 1, 0.15, 0.6), -1);
}

TEST(BatteryMonitorTest, DisabledByZeroThresholds) {
    BatteryMonitor monitor(0.0, 0.0);
    EXPECT_EQ(monitor.update(0.1, 0.0), (BatteryMonitor::Flags{false, false}));
    EXPECT_EQ(monitor.update(0.1, 100.0), (BatteryMonitor::Flags{false, false}));
}

TEST(BatteryMonitorTest, HoldTime) {
    BatteryMonitor monitor(7.0, 6.4, 0.3, 3.0);
    EXPECT_EQ(monitor.update(6.8, 0.0), (BatteryMonitor::Flags{false, false}));
    EXPECT_EQ(monitor.update(6.8, 2.9), (BatteryMonitor::Flags{false, false}));
    EXPECT_EQ(monitor.update(6.8, 3.0), (BatteryMonitor::Flags{true, false}));

    // A dip that recovers before the hold expires never trips.
    BatteryMonitor monitor2(7.0, 6.4, 0.3, 3.0);
    EXPECT_EQ(monitor2.update(6.8, 0.0), (BatteryMonitor::Flags{false, false}));
    EXPECT_EQ(monitor2.update(7.5, 1.0), (BatteryMonitor::Flags{false, false}));
    EXPECT_EQ(monitor2.update(6.8, 2.0), (BatteryMonitor::Flags{false, false}));
    EXPECT_EQ(monitor2.update(6.8, 4.9), (BatteryMonitor::Flags{false, false}));
    EXPECT_EQ(monitor2.update(6.8, 5.0), (BatteryMonitor::Flags{true, false}));
}

TEST(BatteryMonitorTest, HysteresisOnRecovery) {
    BatteryMonitor monitor(7.0, 0.0, 0.3, 0.0);
    EXPECT_EQ(monitor.update(6.9, 0.0), (BatteryMonitor::Flags{true, false}));
    // Inside the hysteresis band the flag stays raised.
    EXPECT_EQ(monitor.update(7.2, 1.0), (BatteryMonitor::Flags{true, false}));
    // Above threshold + hysteresis it clears immediately.
    EXPECT_EQ(monitor.update(7.4, 2.0), (BatteryMonitor::Flags{false, false}));
}

TEST(BatteryMonitorTest, CriticalIndependentOfWarning) {
    BatteryMonitor monitor(7.0, 6.4, 0.3, 0.0);
    EXPECT_EQ(monitor.update(6.9, 0.0), (BatteryMonitor::Flags{true, false}));
    EXPECT_EQ(monitor.update(6.3, 1.0), (BatteryMonitor::Flags{true, true}));
}

// The busy flag is the top of the precedence ladder: it is the one state the
// operator asked for by hand (a front-panel hold — a pairing scan or a
// network-mode switch), so nothing may take the panel while they wait on the
// spinners. The node ORs /bluetooth/scanning and /display/busy into it, which
// is why the field is not named after either topic.
TEST(ExpressionPolicy, BusyWearsTheScanningExpressionGazeCentered) {
    auto in = makeInputs();
    in.busy = true;
    const DisplayTarget target = decide(in, CONFIG, kIdleTarget);
    EXPECT_EQ(target.expression, Expression::SCANNING);
    EXPECT_EQ(target.gaze, GazeDirection::CENTER);
}

TEST(ExpressionPolicy, BusyOutranksEvenACriticalPack) {
    auto in = makeInputs();
    in.busy = true;
    in.battery_critical = true;
    in.battery_low = true;
    EXPECT_EQ(decide(in, CONFIG, kIdleTarget).expression, Expression::SCANNING);
}

TEST(ExpressionPolicy, BusySuppressesFaceAnimations) {
    // A drift animation would fight the spinners for the gaze they own.
    auto in = makeInputs();
    in.busy = true;
    EXPECT_FALSE(selectFaceAnimation(in, CONFIG).has_value());
}
